File: python/common/test_framework.py
class TestFailure(AssertionError):
    """Raised when one or more test cases fail, so runners exit non-zero."""


def run_tests(test_cases):
    """
    Run the provided test cases, print the results, and FAIL LOUDLY on any miss.

    Prints a ✅/❌ line per case and a summary, then raises ``TestFailure`` if any
    case failed. Raising (rather than only printing) is what lets ``run.py`` and
    ``pytest`` detect failures via a non-zero exit / failed assertion.

    :param test_cases: A list of tuples ``(function, args, expected, test_name)``.
    """
    passed = 0
    total = len(test_cases)
    failures = []

    for func, args, expected, name in test_cases:
        result = func(*args)
        if result == expected:
            passed += 1
            print(f"✅  {name}")
        else:
            failures.append(name)
            print(f"❌  {name}")
            print(f"      expected: {expected}")
            print(f"      got:      {result}")

    print("\n" + "─" * 34)
    if not failures:
        print(f"🎉  {passed}/{total} passed — all green")
    else:
        print(f"❌  {passed}/{total} passed · {len(failures)} failed")
        raise TestFailure(f"{len(failures)}/{total} failed: {', '.join(failures)}")

File: javascript/common/test_framework.js
// Minimal test helper — the JavaScript counterpart of the other common/ helpers.
// Solutions load it via NODE_PATH (set by run.py) as:
//   const { runTests } = require("common/test_framework");
// Prints ✅/❌ per case and exits non-zero if any case fails.
function runTests(cases) {
  let passed = 0;
  const failures = [];
  const eq = (a, b) => JSON.stringify(a) === JSON.stringify(b);

  for (const [fn, args, expected, name] of cases) {
    const got = fn(...args);
    if (eq(got, expected)) {
      passed++;
      console.log(`✅ ${name} passed`);
    } else {
      failures.push(name);
      console.log(`❌ ${name} failed`);
      console.log(`   Expected: ${JSON.stringify(expected)}`);
      console.log(`   Got:      ${JSON.stringify(got)}`);
    }
  }

  console.log(`\nTest Results: ${passed}/${cases.length} passed`);
  if (failures.length) {
    console.log(`${failures.length} test(s) failed`);
    process.exit(1);
  }
  console.log("All tests passed! 🎉");
}

module.exports = { runTests };
